### ISSUE TITLE
Add "getEntityManager" or "cleanup" method to the "ManagerInterface"

### ISSUE THREAD

**Issue by vyshkant** [sonata-project/SonataCoreBundle#468]
# The problem
The problem occurred from here: https://github.com/sonata-project/SonataMediaBundle/pull/1259#issuecomment-343722713

In that PR next line was removed:

```php
//clear entity manager for saving memory
$this->getMediaManager()->getEntityManager()->clear();
```

The problem was, that  the `getEntityManager` method does not exist in `Sonata\CoreBundle\Model\ManagerInterface`, which is returned by `$this->getMediaManager()`.

It exists in `Sonata\CoreBundle\Model\BaseEntityManager`, the ancestor of `Sonata\MediaBundle\Entity\MediaManager`.

But `Sonata\CoreBundle\Model\BaseEntityManager` **does not** implement `Sonata\CoreBundle\Model\ManagerInterface` (which is returned by `$this->getMediaManager()`), while its child `Sonata\MediaBundle\Entity\MediaManager` does.

# Possible solutions

I'm not sure if this corresponds to the ideology of the project, but if we want to call `$this->getMediaManager()->getEntityManager()->clear();` on `Sonata\CoreBundle\Model\ManagerInterface`, we have to add the appropriate method to this interface.

We also have to make `Sonata\CoreBundle\Model\BaseEntityManager` implement `Sonata\CoreBundle\Model\ManagerInterface`.

cc @core23

**Comment by greg0ire** [comment]
> we have to add the appropriate method to this interface.

or create a new interface and implement both, if that makes sense.

**Comment by vyshkant** [comment]
> or create a new interface and implement both, if that makes sense.

but we have to be sure, that `$this->getMediaManager()` returns an instance of both, isn't it?

**Comment by greg0ire** [comment]
I don't know... do absolutely all implementation have a clear method? Also, adding new methods to an interface is a BC break, creating a new interface isn't, and you could do this:

```
if ($this->getMediaManager() instanceof ClearableEntityManagerAwareInterface) {
    $this->getMediaManager()->getEntityManager()->clear();
}
```

**Comment by vyshkant** [comment]
> adding new methods to an interface is a BC break

Yes, but this is not a security bug, so it can easily wait for next major release.

**Comment by vyshkant** [comment]
And if we're creating something lile `ClearableEntityManagerAwareInterface`, should it provide `getEntityManager` or `cleanup` method?

**Comment by greg0ire** [comment]
Hmmm... probably `cleanup`, I think using `->` is a code smell (Law of Demeter violation?)

**Comment by vyshkant** [comment]
But I still think that if the only reason of creating a separate `ClearableEntityManagerAwareInterface` is to avoid BC break, than we can just modify `ManagerInterface` in the PR to `master` branch.

**Comment by greg0ire** [comment]
If it makes no sense to have a `ManagerInterface` without a `cleanup` method, I think the right fix is to implement that method in 3.x, and trigger a deprecation error if the `MediaManager` does not have it, and once 3.x is merged into master, add it to the interface in a master PR.

**Comment by core23** [comment]
The classes are deprecated here. Please open a new issue at https://github.com/sonata-project/sonata-doctrine-extensions, if you have still the same problem.
